internal/venice-common/src/main/java/com/linkedin/venice/pubsub/manager/TopicManager.java

```diff
@@ -651,72 +618,110 @@ public synchronized boolean containsTopicAndAllPartitionsAreOnline(
-    return partitionOffsetFetcher.getProducerTimestampOfLastDataRecord(pubSubTopicPartition, retries);
+  public long getLatestOffsetCached(PubSubTopic pubSubTopic, int partitionId) {
+    return topicMetadataFetcher.getLatestOffsetCached(new PubSubTopicPartitionImpl(pubSubTopic, partitionId));
+  }
+
+  public long getProducerTimestampOfLastDataMessageWithRetries(PubSubTopicPartition pubSubTopicPartition, int retries) {
+    return topicMetadataFetcher.getProducerTimestampOfLastDataMessageWithRetries(pubSubTopicPartition, retries);
+  }
+
+  public long getProducerTimestampOfLastDataMessageCached(PubSubTopicPartition pubSubTopicPartition) {
+    return topicMetadataFetcher.getProducerTimestampOfLastDataMessageCached(pubSubTopicPartition);
   }
 
   /**
    * Get offsets for only one partition with a specific timestamp.
    */
-  public long getPartitionOffsetByTime(PubSubTopicPartition pubSubTopicPartition, long timestamp) {
-    return partitionOffsetFetcher.getPartitionOffsetByTime(pubSubTopicPartition, timestamp);
+  public long getOffsetByTime(PubSubTopicPartition pubSubTopicPartition, long timestamp) {
+    return topicMetadataFetcher.getOffsetForTimeWithRetries(pubSubTopicPartition, timestamp, 25);
   }
 
   /**
-   * Get a list of {@link PubSubTopicPartitionInfo} objects for the specified topic.
-   * @param topic
-   * @return
+   * Invalidate the cache for the given topic and all its partitions.
+   * @param pubSubTopic the topic to invalidate
    */
-  public List<PubSubTopicPartitionInfo> partitionsFor(PubSubTopic topic) {
-    return partitionOffsetFetcher.partitionsFor(topic);
+  public CompletableFuture<Void> invalidateCache(PubSubTopic pubSubTopic) {
+    return topicMetadataFetcher.invalidateKeyAsync(pubSubTopic);
   }
 
-  public String getPubSubBootstrapServers() {
-    return this.pubSubBootstrapServers;
+  /**
+   * Prefetch and cache the latest offset for the given topic-partition.
+   * @param pubSubTopicPartition the topic-partition to prefetch and cache the latest offset for
+   */
+  public void prefetchAndCacheLatestOffset(PubSubTopicPartition pubSubTopicPartition) {
+    topicMetadataFetcher.populateCacheWithLatestOffset(pubSubTopicPartition);
   }
 
-  @Override
-  public synchronized void close() {
-    Utils.closeQuietlyWithErrorLogged(partitionOffsetFetcher);
-    pubSubReadOnlyAdminAdapter.ifPresent(Utils::closeQuietlyWithErrorLogged);
-    pubSubWriteOnlyAdminAdapter.ifPresent(Utils::closeQuietlyWithErrorLogged);
+  public String getPubSubClusterAddress() {
+    return this.pubSubClusterAddress;
   }
 
-  // For testing only
-  public void setTopicConfigCache(Cache<PubSubTopic, PubSubTopicConfiguration> topicConfigCache) {
-    this.topicConfigCache = topicConfigCache;
+  @Override
+  public void close() {
+    if (isClosed.get()) {
+      logger.warn("{} is already closed", this);
+    }
+    CompletableFuture.runAsync(() -> {
```

Comment:
Why closing it asynchronously?

Reply:
We want TM close to be best effort operation and don't want to block for it to complete

Reply:
So is this a new logic or existing behavior?

Reply:
New logic

Reply:
Got it.
So you worry that TopicManager#close can take a long time and it might block the stop procedure, right?
Today, I think most of the `close` functions are doing sync logic and if we want to speed it up, we can execute the closing logic concurrently if there are multiple closables.
If we can make it sync, it would be easier to reason IMO.

Reply:
Yeah initially TM::close was sync and we were closing all TMs in TMR concurrently. But I removed it after discussion with @FelixGV: we decided waiting for it to complete doesn't add significant value.

Here is the commit that removed sync close: https://github.com/linkedin/venice/pull/743/commits/70be6aa13f323a5edbbb557b02f6999d178f4d8d

Reply:
Got it.
If the impact has been discussed extensively, I trusted your conclusion.